### PR TITLE
add wgan-graphs.py tutorial

### PR DESCRIPTION
**Pull request by akensert**
A generative model for small graphs (specifically small molecular graphs).

### CODE REVIEW

File: examples/graph/wgan-graphs.py [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+[SMILES](https://en.wikipedia.org/wiki/Simplified_molecular-input_line-entry_system) to
+molecule objects, and then from those obtain sets of atoms and bonds.
+
+Notice, RDKit is commonly installed via [Conda](https://www.rdkit.org/docs/Install.html).
+However, thanks to [kuelumbus'](https://github.com/kuelumbus/)
+[rdkit_platform_wheels](https://github.com/kuelumbus/rdkit_platform_wheels), rdkit
+can now (for the sake of this tutorial) be installed easily via pip.
+"""
+
+"""shell
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:44]: Rather than including this block, list the required dependencies and show a code snippet of the pip commands. Readers should be able to choose to install them.

File: examples/graph/wgan-graphs.py [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+import tensorflow as tf
+from tensorflow import keras
+import numpy as np
+
+RDLogger.DisableLog("rdApp.*")
+
+"""
+## Dataset
+
+The dataset used in this tutorial is a [quantum mechanics
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:65]: Make sure all markdown links fit in one line (otherwise they won't render)

File: examples/graph/wgan-graphs.py [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+dataset](http://quantum-machine.org/datasets/) (QM9), obtained from
+[MoleculeNet](http://moleculenet.ai/datasets-1). Although many feature and label columns
+come with the dataset, we'll only focus on the
+[SMILES](https://en.wikipedia.org/wiki/Simplified_molecular-input_line-entry_system)
+column ("smiles"). The QM9 dataset is a good first dataset to work with for generating
+graphs, as the maximum number of \[heavy\] atoms found in a molecule is only nine.
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:71]: What's \[heavy\]? A label?

File: examples/graph/wgan-graphs.py [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+RDLogger.DisableLog("rdApp.*")
+
+"""
+## Dataset
+
+The dataset used in this tutorial is a [quantum mechanics
+dataset](http://quantum-machine.org/datasets/) (QM9), obtained from
+[MoleculeNet](http://moleculenet.ai/datasets-1). Although many feature and label columns
+come with the dataset, we'll only focus on the
+[SMILES](https://en.wikipedia.org/wiki/Simplified_molecular-input_line-entry_system)
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:69]: Please explain what that is.

File: examples/graph/wgan-graphs.py [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+BOND_DIM = 4 + 1  # Number of bond types
+LATENT_DIM = 64  # Size of the latent space
+
+
+def smiles_to_graph(smiles):
+    # Converts SMILES to molecule object
+    mol = Chem.MolFromSmiles(smiles)
+
+    # Initialize adjacency and feature tensor
+    A = np.zeros((BOND_DIM, NUM_ATOMS, NUM_ATOMS), "float32")
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:136]: Always use fully-spelled out variable names, to improve readability.

File: examples/graph/wgan-graphs.py [hunk 6]
```diff
@@ -0,0 +1,545 @@
+"""
+Title: WGAN-GP with R-GCN for the generation of small molecular graphs
+Author: [akensert](https://github.com/akensert)
+Date created: 2021/06/27
+Last modified: 2021/06/27
+Description: Complete implementation of WGAN-GP with R-GCN to generate novel molecules.
+"""
+
+"""
+## References
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:10]: Start with an introduction describing what the tutorial teaches you and why it's important. References can go at the end of the introduction.

File: examples/graph/wgan-graphs.py [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+    feature_tensor.append(X)
+
+adjacency_tensor = np.array(adjacency_tensor)
+feature_tensor = np.array(feature_tensor)
+
+print("A.shape =", adjacency_tensor.shape)
+print("H.shape =", feature_tensor.shape)
+
+"""
+## Model
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:221]: Provide details/explanations about the  modeling choices you make.

File: examples/graph/wgan-graphs.py [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+    x = z
+    for units in dense_units:
+        x = keras.layers.Dense(units, activation="tanh")(x)
+        x = keras.layers.Dropout(dropout_rate)(x)
+
+    # Map outputs of previous layers (x) to [continuous] adjacency tensors (x_A)
+    x_A = keras.layers.Dense(tf.math.reduce_prod(adjacency_shape))(x)
+    x_A = keras.layers.Reshape(adjacency_shape)(x_A)
+    # Symmetrify tensors in the last two dimensions
+    x_A = keras.layers.Lambda(lambda x: (x + tf.transpose(x, (0, 1, 3, 2))) / 2)(x_A)
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:242]: You can just use `tf.transpose` directly, no lambda.

File: examples/graph/wgan-graphs.py [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+
+        # Generate interpolated graphs (A_interp and H_interp)
+        alpha = tf.random.uniform([self.batch_size])
+        alpha = tf.reshape(alpha, (self.batch_size, 1, 1, 1))
+        A_interp = (A_real * alpha) + (1 - alpha) * A_gen
+        alpha = tf.reshape(alpha, (self.batch_size, 1, 1))
+        H_interp = (H_real * alpha) + (1 - alpha) * H_gen
+
+        # Compute the logits of interpolated graphs
+        with tf.GradientTape() as tape:
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:478]: This gradient tape appears to be nested, in this intended?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! This looks very interesting.

File: examples/graph/wgan-graphs.py (hunk 9 above)
Comment (akensert) [reply examples/graph/wgan-graphs.py:525]: As far as I understand: Yes :)

File: examples/graph/wgan-graphs.py (hunk 3 above)
Comment (akensert) [reply examples/graph/wgan-graphs.py]: Non-hydrogen atoms.

**Comment by akensert** [comment]
Thanks for the feedback! Much appreciated. In the new commit, all issues should be fixed :)

File: examples/graph/wgan-graphs.py (hunk 3 above)
Comment (akensert) [reply examples/graph/wgan-graphs.py]: Notice, hydrogen atoms are usually excluded to reduce the size of the tensors/matrices. This should also make it easier to train a model. Hydrogen atoms can be inferred by RDKit later. I'll make another commit clarifying this briefly.

File: examples/graph/wgan-graphs.py [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+can alleviate the search for good candidate drugs, by predicting properties of known molecules
+(e.g., solubility, toxicity, affinity to target protein, etc.). As the number of
+possible molecules is astronomical, the space in which we search for/explore molecules is
+just a fraction of the entire space. Therefore, it's arguably desirable to implement
+generative models that can learn to generate novel molecules (which would otherwise have never been explored).
+
+### References (implementation)
+
+The implementation in this tutorial is based on/inspired by the [MolGAN
+paper](https://arxiv.org/abs/1805.11973) and DeepChem's [Basic
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:26]: Make sure markdown links fit on one line.

File: examples/graph/wgan-graphs.py [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+MolGAN](https://deepchem.readthedocs.io/en/latest/api_reference/models.html#basicmolganmod
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:27]: Here too.

File: examples/graph/wgan-graphs.py [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+human-readable. Encoding molecules as a string both alleviates and facilitates database
+and/or web searching of a given molecule. RDKit uses algorithms to
+accurately transform a given SMILES to a molecule object, which can then
+be used to compute a great number of molecular properties/features.
+
+Notice, RDKit is commonly installed via [Conda](https://www.rdkit.org/docs/Install.html).
+However, thanks to
+[rdkit_platform_wheels](https://github.com/kuelumbus/rdkit_platform_wheels), rdkit
+can now (for the sake of this tutorial) be installed easily via pip, as follows:
+```python
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:62]: Remove "python", that's shell.

File: examples/graph/wgan-graphs.py [hunk 13]
```diff
@@ -0,0 +1,601 @@
...
+pip -q install rdkit-pypi
+```
+And to allow easy visualization of a molecule objects, Pillow needs to be installed:
+```python
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:66]: Here too.

File: examples/graph/wgan-graphs.py [hunk 14]
```diff
@@ -0,0 +1,601 @@
...
+
+The generator network needs to be able to map (for each example in the batch) a vector `z`
+to a 3-D adjacency tensor (`A`) and 2-D feature tensor (`H`). For this, `z` will first be
+passed through a fully-connected network, for which the output will be further passed
+through two separate fully-connected networks. Each of these two fully-connected
+networks will then output (for each example in the batch) a tanh-activated vector
+followed by a reshape and softmax to match that of a multi-dimensional adjacency/feature
+tensor.
+
+As the discriminator network will recieves as input a graph (A, H) from either the
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:260]: Use backticks for A, H

File: examples/graph/wgan-graphs.py [hunk 15]
```diff
@@ -0,0 +1,601 @@
...
+        self.kernel_initializer = keras.initializers.get(kernel_initializer)
+        self.bias_initializer = keras.initializers.get(bias_initializer)
+        self.kernel_regularizer = keras.regularizers.get(kernel_regularizer)
+        self.bias_regularizer = keras.regularizers.get(bias_regularizer)
+
+    def build(self, input_shape):
+        bond_dim = input_shape[0][1]
+        atom_dim = input_shape[1][2]
+
+        self.W = self.add_weight(
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:360]: Just "w". Alternatively you can use the standard Keras convention `self.kernel` / `self.bias`

File: examples/graph/wgan-graphs.py [hunk 16]
```diff
@@ -0,0 +1,601 @@
...
+        units=128,
+        activation="relu",
+        use_bias=False,
+        kernel_initializer="glorot_uniform",
+        bias_initializer="zeros",
+        kernel_regularizer=None,
+        bias_regularizer=None,
+        **kwargs
+    ):
+        super(RelationalGraphConvLayer, self).__init__(**kwargs)
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:346]: Every time you call `super` in this example, use the Python convention, `super().__init__(**kwargs)`

File: examples/graph/wgan-graphs.py [hunk 17]
```diff
@@ -0,0 +1,601 @@
...
+Author: [akensert](https://github.com/akensert)
+Date created: 2021/06/28
+Last modified: 2021/06/28
+Description: Complete implementation of WGAN-GP with R-GCN to generate novel molecules.
+"""
+
+"""
+## Introduction
+
+In this tutorial, a generative model for graphs will be implemented to generate
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:12]: Prefer using the active voice: "we implement a generative model for graphs and use it to generate new molecule."

File: examples/graph/wgan-graphs.py [hunk 18]
```diff
@@ -0,0 +1,601 @@
...
+wgan.fit([adjacency_tensor, feature_tensor], epochs=10, batch_size=16)
+
+"""
+## Sample novel molecules with the generator
+"""
+
+
+def sample(generator, batch_size):
+    z = tf.random.normal((batch_size, LATENT_DIM))
+    graph = generator(z)
```
Comment (fchollet) [inline examples/graph/wgan-graphs.py:564]: A slight concern here is that `generator(z)` is running eagerly.

Consider using `generator.predict(z)`, or alernatively move the TF part of this function to a `@tf.function`.

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

**Comment by akensert** [comment]
Thanks again for the feedback! Committed another update.

File: examples/graph/wgan-graphs.py (hunk 16 above)
Comment (akensert) [reply examples/graph/wgan-graphs.py]: Notice, I modified the super calls (as per your suggestion) for the other classes too. Not sure if that's what you asked for.

**Review by fchollet — APPROVED** [review]
LGTM, thank you! Please add the generated files.

**Comment by akensert** [comment]
My pleasure! Generated files have now been added.

**Comment by fchollet** [comment]
Thanks for the great contribution! 👍
